bridge/gitlab/import.go

```diff
@@ -35,49 +34,83 @@ func (gi *gitlabImporter) Init(conf core.Configuration) error {
 
 // ImportAll iterate over all the configured repository issues (notes) and ensure the creation
 // of the missing issues / comments / label events / title changes ...
-func (gi *gitlabImporter) ImportAll(repo *cache.RepoCache, since time.Time) error {
-	gi.iterator = NewIterator(gi.conf[keyProjectID], gi.conf[keyToken], since)
-
-	// Loop over all matching issues
-	for gi.iterator.NextIssue() {
-		issue := gi.iterator.IssueValue()
-		fmt.Printf("importing issue: %v\n", issue.Title)
-
-		// create issue
-		b, err := gi.ensureIssue(repo, issue)
-		if err != nil {
-			return fmt.Errorf("issue creation: %v", err)
-		}
-
-		// Loop over all notes
-		for gi.iterator.NextNote() {
-			note := gi.iterator.NoteValue()
-			if err := gi.ensureNote(repo, b, note); err != nil {
-				return fmt.Errorf("note creation: %v", err)
-			}
-		}
-
-		// Loop over all label events
-		for gi.iterator.NextLabelEvent() {
-			labelEvent := gi.iterator.LabelEventValue()
-			if err := gi.ensureLabelEvent(repo, b, labelEvent); err != nil {
-				return fmt.Errorf("label event creation: %v", err)
+func (gi *gitlabImporter) ImportAll(ctx context.Context, repo *cache.RepoCache, since time.Time) (<-chan core.ImportResult, error) {
+	gi.iterator = NewIterator(ctx, 10, gi.conf[keyProjectID], gi.conf[keyToken], since)
+	out := make(chan core.ImportResult)
+	gi.out = out
+
+	go func() {
+		defer close(gi.out)
+
+		// Loop over all matching issues
+		for gi.iterator.NextIssue() {
+			issue := gi.iterator.IssueValue()
+			var b *cache.BugCache
+			var err error
+
+			select {
+			case <-ctx.Done():
+				// if the context is cancelled commit the latest changes
+				if err := b.CommitAsNeeded(); err != nil {
+					err := errors.Wrap(err, "bug commit")
+					out <- core.NewImportError(err, b.Id())
+				}
+
+				out <- core.NewImportError(ctx.Err(), "")
+				return
+			default:
+
+				// create issue
+				b, err = gi.ensureIssue(repo, issue)
+				if err != nil {
+					err := fmt.Errorf("issue creation: %v", err)
+					out <- core.NewImportError(err, "")
+					return
+				}
+
+				// Loop over all notes
+				for gi.iterator.NextNote() {
+					note := gi.iterator.NoteValue()
+					if err := gi.ensureNote(repo, b, note); err != nil {
+						err := fmt.Errorf("note creation: %v", err)
+						out <- core.NewImportError(err, entity.Id(strconv.Itoa(note.ID)))
+						return
+					}
+				}
+
+				// Loop over all label events
+				for gi.iterator.NextLabelEvent() {
+					labelEvent := gi.iterator.LabelEventValue()
+					if err := gi.ensureLabelEvent(repo, b, labelEvent); err != nil {
+						err := fmt.Errorf("label event creation: %v", err)
+						out <- core.NewImportError(err, entity.Id(strconv.Itoa(labelEvent.ID)))
```

Comment:
I have a lot of these transformation in PR (`string` to `entity.Id` and `int` to `entity.Id`) would it be better to add two new functions to the package `entity` ( `IdFromString` and `IdFromInt` ) ?